pkg/credentialprovider/aws/aws_credentials_test.go

```diff
@@ -52,112 +67,221 @@ func (p *testTokenGetter) GetAuthorizationToken(input *ecr.GetAuthorizationToken
+	}
+	if len(creds2) != 2 {
+		t.Errorf("Got more hits than expected: %s", creds2)
+	}
+
+	if creds2[0].Password != creds2[1].Password {
+		t.Errorf("cached credentials do not match")
+	}
+	if creds1[0].Password != creds2[0].Password {
+		t.Errorf("cached credentials do not match")
+	}
+}
+
+func TestChinaECRProvide(t *testing.T) {
 	registry := "123456789012.dkr.ecr.cn-foo-1.amazonaws.com.cn"
 	otherRegistries := []string{
 		"123456789012.dkr.ecr.lala-land-1.amazonaws.com",
 		"private.registry.com",
 		"gcr.io",
 	}
-	image := "foo/bar"
-
-	provider := newEcrProvider("cn-foo-1",
-		&testTokenGetter{
+	image := path.Join(registry, "foo/bar")
+	p := newECRProvider(&testTokenGetterFactory{
+		getter: &testTokenGetter{
 			user:     user,
 			password: password,
 			endpoint: registry,
-		})
-
+		},
+	})
 	keyring := &credentialprovider.BasicDockerKeyring{}
-	keyring.Add(provider.Provide())
-
+	keyring.Add(p.Provide(image))
 	// Verify that we get the expected username/password combo for
 	// an ECR image name.
-	fullImage := path.Join(registry, image)
-	creds, ok := keyring.Lookup(fullImage)
+	creds, ok := keyring.Lookup(image)
 	if !ok {
-		t.Errorf("Didn't find expected URL: %s", fullImage)
+		t.Errorf("Didn't find expected URL: %s", image)
 		return
 	}
 	if len(creds) > 1 {
 		t.Errorf("Got more hits than expected: %s", creds)
 	}
-	val := creds[0]
-
-	if user != val.Username {
-		t.Errorf("Unexpected username value, want: _token, got: %s", val.Username)
+	cred := creds[0]
+	if user != cred.Username {
+		t.Errorf("Unexpected username value, want: %s, got: %s", user, cred.Username)
 	}
-	if password != val.Password {
-		t.Errorf("Unexpected password value, want: %s, got: %s", password, val.Password)
+	if password != cred.Password {
+		t.Errorf("Unexpected password value, want: %s, got: %s", password, cred.Password)
 	}
-	if email != val.Email {
-		t.Errorf("Unexpected email value, want: %s, got: %s", email, val.Email)
+	if email != cred.Email {
+		t.Errorf("Unexpected email value, want: %s, got: %s", email, cred.Email)
 	}
 
 	// Verify that we get an error for other images.
 	for _, otherRegistry := range otherRegistries {
-		fullImage = path.Join(otherRegistry, image)
-		creds, ok = keyring.Lookup(fullImage)
+		image = path.Join(otherRegistry, image)
+		creds, ok = keyring.Lookup(image)
 		if ok {
-			t.Errorf("Unexpectedly found image: %s", fullImage)
+			t.Errorf("Unexpectedly found image: %s", image)
 			return
 		}
 	}
 }
+
+func TestChinaECRProvideCached(t *testing.T) {
+	registry := "123456789012.dkr.ecr.cn-foo-1.amazonaws.com.cn"
+	p := newECRProvider(&testTokenGetterFactory{
+		getter: &testTokenGetter{
+			user:              user,
+			password:          password,
+			endpoint:          registry,
+			randomizePassword: true,
+		},
+	})
+	image := path.Join(registry, "foo/bar")
+	keyring := &credentialprovider.BasicDockerKeyring{}
+	keyring.Add(p.Provide(image))
+	// time.Sleep(6 * time.Second) //for testing with the cache expiring
```

Comment:
You can use a fake clock if you actually want to test cache expiration, but that might be a good thing to do in another PR as it'll likely make this one much bigger

Reply:
SGTM. Won't do it here.